### ISSUE TITLE
Rule suggestion: no direct key lookups, prefer wrappers

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#1879]
Key lookups in TypeScript return the incorrect type, e.g.

``` ts
const xs = [1];
xs[500] // number, should be number | undefined
```

Therefore, it is good practice to avoid using key lookups, and instead always prefer helper functions:

``` ts
const getInArray = <T>(array: T[], index: number): T | undefined => array[index]
getInArray(xs, 500) // number | undefined
```

The same is true for objects, e.g. `foo[bar]`.

As the TypeScript team do not want to fix this, it would be great if TSLint had a rule to enforce this.

In my experience, a lot of bugs come from unguarded key lookups like this.

**Comment by JoshuaKGoldberg** [comment]
@OliverJAsh this seems like it would be very disruptive. If enabled, it would basically ban direct uses of `[]` lookups, right? 

-1 from me, but curious if there's a lot of desire for this.

**Comment by JoshuaKGoldberg** [comment]
Ha, always nice to see my opinions aligning with the TypeScript team. 😆

To recap, as RyanCavanaugh mentioned and you responded to, the problem is that this rule would flag _all_ array lookups. It's hard to tell if this would help more than hurt the majority of users, which means it's hard to tell if this should be added to TSLint core. 

Putting it another way: if we find that this hurts more than it helps, we're then stuck with maintaining this obscure rule for years (see `no-use-before-declare`). 

Proposal: can you make a separate repository + npm package, publish the rule there, and ask users to try it out? That would let you iterate more quickly than sending PRs to this repository and waiting for TSLint releases to flight updates.

**Comment by adidahiya** [comment]
what is considered "iterating an enumerated collection"? is it `Array.prototype.forEach` and for-of loops?

this still feels like a niche rule which I am reluctant to bring into tslint core. awaiting more community feedback.

**Comment by JoshuaKGoldberg** [comment]
Closing for inactivity. Thanks all!
